qcfractal/storage_sockets/sqlalchemy_common.py

```diff
@@ -0,0 +1,192 @@
+        else:
+            # Existing record found
+            updated_idx.append(idx)
+
+            # Do update. These changes are propagated to the database on commit
+            for c in update_cols:
+                new_value = getattr(rec, c)
+                setattr(existing[0], c, new_value)
+
+            ret.append(existing[0])
+
+    session.commit()
+
+    meta = UpsertMetadata(inserted_idx=inserted_idx, updated_idx=updated_idx, errors=errors)
+    return meta, ret
+
+
+def _insert_general(
+    session: "sqlalchemy.orm.session.Session", data: Sequence[ORM_T], search_cols: Iterable[str]
+) -> Tuple[UpsertMetadata, List[Union[ORM_T, None]]]:
+    """
+    Perform a general insert, taking into account existing data
+
+    For each ORM object in data, check if the object/row already exists in the database. If it doesn't exist,
+    add it to the database. If the row does exist, the existing record will be returned.
+    to the database. The columns passed to search_cols will be used to determine if the data/row already exists.
+
+    If the row does not exist, but the input record has an auto-incremented primary key set, then
+    that is considered an error.
+
+    A list of ORM objects is returned. The order is the same as was given in the data list. These will
+    correspond to rows in the database and will have primary keys, etc, filled in. These may alias variables
+    in the data parameter. If an error occurs for a particular record, that index in the returned data will be None.
+
+    The returned ORM objects are attached to the given session, but the records are committed
+
+    Objects contained in the data parameter may be modified.
+
+
+    Parameters
+    ----------
+    session: sqlalchemy.orm.session.Session
+        An existing SQLAlchemy session to use for adding/updating
+    data: Sequence[ORM_T]
+        List/Sequence of ORM objects to be added to the database. These objects may be modified in-place.
+    search_cols: Iterable[str]
+        What columns to use to determine if data already exists in the database.
+
+    Returns
+    -------
+    Tuple[InsertMetadata, List[Union[Base, None]]
+        Metadata showing what was added/updated, and a list of ORM objects. These
+        ORM objects reflect existing or automatically-generated database fields (for example, ids).
+        If an error occurs, the corresponding entry in the list will be None.
+    """
+
+    # These are the records/ORM objects to be returned
+    ret = []
+
+    # Indices of the returned list corresponding to inserted and existing records
+    inserted_idx = []
+    existing_idx = []
+    errors = []
+
+    # This is the type of the ORM. It should be the same for all items in the list
+    orm_type = type(data[0])
+    assert all(type(x) == orm_type for x in data)
+
+    for idx, rec in enumerate(data):
+        # Build up the filter dictionary
+        query = {}
+        for c in search_cols:
+            query[c] = getattr(rec, c)
```

Comment:
Since you have the keys already you can likely generalize the bulk add by building a key tuple (assuming `search_col` order is static) that you can map back and forth between objects and indices. A single query can then be performed to check of existing and then a bulk add performed which will be ~100-1000x faster than current.

I'm unsure of your current performance constraints, but worth thinking about if needed.